web-server/v0.4/src/pages/Dashboard/Controllers.js

```diff
@@ -27,7 +27,13 @@ export default class Controllers extends Component {
   }
 
   componentDidMount() {
-    this.queryDatastoreConfig();
+    const { controllers, selectedIndices, indices } = this.props;
+
+    if (controllers.length === 0 || selectedIndices.length === 0 || indices.length === 0) {
+      this.queryDatastoreConfig();
+    } else {
+      this.setState({ controllers });
```

Comment:
Why do we set the state to the same state it was here?  If we need this, can you add a comment explaining why it is necessary?

Reply:
We set controllers to the state here in the case where the user has spun up the dashboard session again because the component needs to remount. Since the data has persisted from the last session, we don't need to make the necessary queries to initialize the view; however, our Table component in this page component depends on accessing controllers from the state rather than the props for search manipulation.

Reply:
Okay, that seems suspect.  Do we know the user has spun up the dashboard again?  Or are we just doing it because it "could" be the case they have spun it up again?

If we know for sure, that is one thing, but if this is just to cover the "could" case, can't we use additional data from the environment to tell for sure?

Reply:
We know the user has spun up the dashboard again or for the first time because this logic is executed in the `componentDidMount()` method which is invoked after a component is mounted to the DOM tree. Since the `Controllers` component is an entry point and defined as the base url `/dashboard`, we can check the following:

1. the `Controllers` component has remounted (logic within `componentDidMount()` lifecycle method is referenced)
2. the necessary component props data has persisted when restarting the session

Reply:
I'm including a cleaner solution for handling the handoff from the controllers prop to the controllers state data. Essentially, we'll initialize the state data with the controllers prop in the constructor of the component and use `componentWillReceiveProps()` to make necessary updates to the state data when the controllers prop changes. This way we avoid update the state in `componentDidMount()` and remove an "extra check" we have to perform during the persisted lifecycle.